index/segment/mem/segment.go

```diff
@@ -93,24 +95,38 @@ func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 
 func (s *segment) Insert(d doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
 		return sgmt.ErrClosed
 	}
 
 	if s.state.sealed {
-		s.state.RUnlock()
 		return errSegmentSealed
 	}
 
-	// Validate that the document contains only valid UTF-8.
-	for _, f := range d.Fields {
-		if !utf8.Valid(f.Name) {
-			return fmt.Errorf("document contains invalid field name: %v", f.Name)
+	docID, err := d.Validate()
```

Comment:
could you also expose the path that skips validation prior to insert?

Reply:
As in have an option to not validate?

Reply:
leave for now